sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java

```diff
@@ -0,0 +1,909 @@
+      super(context, o11y, backoff, sleeper);
+    }
+
+    @Override
+    public void recordRequestStart(Instant start) {
+      at.recordStartRequest(start);
+      this.start = start;
+      state = AttemptState.STARTED;
+    }
+
+    @Override
+    public void recordStreamValue(Instant now) {
+      state.checkActive();
+      o11y.rpcStreamValueReceived.inc();
+    }
+  }
+
+  final class RpcWriteAttemptImpl extends BaseRpcAttempt implements RpcWriteAttempt {
+
+    private RpcWriteAttemptImpl(Context context, O11y o11y, BackOff backoff, Sleeper sleeper) {
+      super(context, o11y, backoff, sleeper);
+    }
+
+    @Override
+    public boolean awaitSafeToProceed(Instant instant) throws InterruptedException {
+      state.checkActive();
+      Optional<Duration> shouldThrottle = writeRampUp.shouldThrottle(instant);
+      if (shouldThrottle.isPresent()) {
+        Duration throttleDuration = shouldThrottle.get();
+        long throttleDurationMillis = throttleDuration.getMillis();
+        getLogger().debug("Still ramping up, Delaying request by {}ms", throttleDurationMillis);
+        throttleRequest(throttleDuration);
+        return false;
+      } else {
+        return super.awaitSafeToProceed(instant);
+      }
+    }
+
+    @Override
+    public <T, ElementT extends Element<T>> FlushBufferImpl<T, ElementT> newFlushBuffer(
+        Instant instantSinceEpoch) {
+      state.checkActive();
+      int availableWriteCountBudget = writeRampUp.getAvailableWriteCountBudget(instantSinceEpoch);
+      int nextBatchMaxCount = wb.nextBatchMaxCount(instantSinceEpoch);
+      int batchMaxCount =
+          Ints.min(
+              Math.max(0, availableWriteCountBudget),
+              Math.max(0, nextBatchMaxCount),
+              options.getBatchMaxCount());
+      o11y.batchCapacityCount.update(batchMaxCount);
+      return new FlushBufferImpl<>(batchMaxCount, options.getBatchMaxBytes());
+    }
+
+    @Override
+    public void recordRequestStart(Instant start, int numWrites) {
+      at.recordStartRequest(start, numWrites);
+      writeRampUp.recordWriteCount(start, numWrites);
+      this.start = start;
+      state = AttemptState.STARTED;
+    }
+
+    @Override
+    public void recordWriteCounts(Instant end, int successfulWrites, int failedWrites) {
+      int totalWrites = successfulWrites + failedWrites;
+      state.checkStarted();
+      wb.recordRequestLatency(start, end, totalWrites, o11y.latencyPerDocumentMs);
+      if (successfulWrites > 0) {
+        at.recordSuccessfulRequest(start, successfulWrites);
+      }
+      if (failedWrites > 0) {
+        at.recordFailedRequest(start, failedWrites);
+      }
+    }
+  }
+
+  /**
+   * Determines batch sizes based on past performance.
+   *
+   * <p>It aims for a target response time per RPC: it uses the response times for previous RPCs and
+   * the number of documents contained in them, calculates a rolling average time-per-document, and
```

Comment:
Correct, right now it is only tracking number of documents not the size of those documents. It could be extended if we felt so inclined to do so at some point.